### DISCUSSION TITLE
How to update the frontmatter to add a complete date in frontmatter in a finished?

### DISCUSSION THREAD

**Discussion by francisHugo** [SilentVoid13/Templater#1242] [Help] [answered]
hello, 

I am trying to add a complete date when I finish a note, which already have some frontmatter in it. for example, when I finished a note like 

```
---
created by: Tom
---

## Some words
more text

```

I found a template snippet which can add `status: In progress` and `review count: .nan``. After I used the snippets, the note became like this:

```
---
created by: Tom
status: In progress
review count: .nan
---

## Some words
more text

```
## The snippets I found from https://zachyoung.dev/posts/templater-snippets#suggester-for-files-in-a-specific-folder is like below:
```
<%*
const file = tp.file.find_tfile(tp.file.path(true));
await app.fileManager.processFrontMatter(file, (frontmatter) => {
  frontmatter["status"] = "In progress";
  frontmatter["review count"] += 1;
  delete frontmatter["ignored"];
})
-%>
```
## Questions:

1. Is there any way to add a `completed date: 2023-11-13 ` ?
2. if the finished note does not have a ` review count `, can set the review count to 0? other than .nan?

I know a very little template, I made some changes, but it does not work. Please help. Thanks.

## What I made: (does not work :-( )
```
<%*
const file = tp.file.find_tfile(tp.file.title);
await app.fileManager.processFrontMatter(file, (frontmatter) => {
frontmatter["status"] = "In progress";
frontmatter["reviewTimes"] += 1;
delete frontmatter["ignored"];
frontmatter["reviewedDate"] = "<% tp.date.now("YYYY-MM-DD") %>";
})
-%>
```
Honesty, I do not understand the meaning of `delete frontmatter["ignored"]` .

Thanks in advance,

Sincerely,

FH

**Comment by Zachatoo** [comment]
Howdy!

In it's simplest form, in the `processFrontMatter` [callback function](https://developer.mozilla.org/en-US/docs/Glossary/Callback_function), you are modifying a [JavaScript Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object). Then Obsidian takes that object, converts it to YAML, and replaces the frontmatter in that note with that new YAML.

If you want to set the value for a frontmatter key or add a new key/value pair, you use the `=` operator. If you want to delete the key/value pair you use the [delete](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete) operator.

```js
<%*
const file = tp.file.find_tfile(tp.file.path(true));
await app.fileManager.processFrontMatter(file, (frontmatter) => {
  // Add or set the value
  frontmatter["key"] = "value";
  // Delete a key/value pair
  delete frontmatter["key"];
});
-%>
```

---

> Is there any way to add a completed date: `2023-11-13`?

Assuming you want it to be "today", you can use `tp.date.now()` in `processFrontMatter` like this (notice how I don't nest `<% ... <% ... %> ... %>`):

```js
<%*
const file = tp.file.find_tfile(tp.file.path(true));
await app.fileManager.processFrontMatter(file, (frontmatter) => {
  frontmatter["completed date"] = tp.date.now("YYYY-MM-DD");
});
-%>
```

> if the finished note does not have a review count, can set the review count to 0? other than .nan?

Yes! We can use an [if else](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/if...else) statement to conditionally set `review count` to `0` if it doesn't exist.

```js
<%*
const file = tp.file.find_tfile(tp.file.path(true));
await app.fileManager.processFrontMatter(file, (frontmatter) => {
  if (frontmatter["review count"] === undefined) {
    frontmatter["review count"] = 0;
  } else {
    frontmatter["review count"] += 1;
  }
});
-%>
```

**Comment by francisHugo** [comment]
Many thanks @Zachatoo 

With the help, I made the templater as bellow, the date works well, and the review count( I made it to reviewedTimes) works well, too.  Now, I would like to add an if clause to judge `noteStatus`, if its value is `in progress`, it should add `reviewedDate: date value`, `reviewedTimes +1`, and if its value is `done`, it should add `finishedDate: date value`, `reviewedTimes: +1`, and if I manually change the `noteStatus` from `done` to `in progress` or something else, it should add `reviewedDate`  and delete the `finishedDate` if it exists. 

unfortunately, it does not work. Will you please have look at it and correct it, to make it works? Many thanks.

The templater is shown below:

``` 
<%*
const head = `
`;

const tf = tp.file.find_tfile(tp.file.title);
await app.vault.modify(tf, tp.file.content + head);
-%>


<%*
const file = tp.file.find_tfile(tp.file.title);
await app.fileManager.processFrontMatter(file, (frontmatter) => {
  frontmatter["modifiedBy"] = "no body";
  frontmatter["reviewedBy"] = "no body";
  frontmatter["noteStatus"] = "in progress"
   if (frontmatter["noteStatus"] = "done") {
	    frontmatter["finishedDate"] = tp.date.now("YYYY-MM-DD");
	    delete frontmatter["reviewedDate"];
	  } else if (frontmatter["noteStatus"] = "in progress"){
		frontmatter["reviewedDate"] = tp.date.now("YYYY-MM-DD");
		delete frontmatter["finishedDate"];
	} else {
	frontmatter["reviewedDate"] = tp.date.now("YYYY-MM-DD");
	delete frontmatter["finishedDate"];
	};
	if (frontmatter["reviewedTimes"] === undefined) {
    frontmatter["reviewedTimes"] = 1;
  } else {
    frontmatter["reviewedTimes"] += 1;
  };
})
-%>

```
BTW, after implying the templater, I have to use clt+Z to make the Yaml shown. I really do not know why.

Best Regards,

FH

**Reply by Zachatoo** [reply]
1. Many of your conditions in your `if` statements are using `=` instead of `===` for comparing, which is likely causing unexpected results.
2. You're setting `frontmatter["noteStatus"] = "in progress"` every time this script is run, so when you go to your conditions on the next line, it'll always run the same condition.

With the second point in mind, you may want to consider having two separate templates, one for when you want to review it, and one for when you want to mark it as done? If you're going to be setting the `noteStatus` of your frontmatter, I think you'll have to go with two separate templates, or use `tp.system.prompt` to ask what action you want to take.

**Reply by Zachatoo** [reply]
I'm guessing this top section is what is causing you to have to undo to see the changes. But I'm not 100% sure. I don't really know what this is trying to do.

```js
<%*
const head = `
`;

const tf = tp.file.find_tfile(tp.file.title);
await app.vault.modify(tf, tp.file.content + head);
-%>
```

Might be worth looking into the [tp.hooks](https://silentvoid13.github.io/Templater/internal-functions/internal-modules/hooks-module.html) module to help with not having to undo as well.

**Reply by francisHugo** [reply]
Thanks @Zachatoo ,

Now I understood why my codes always result the same condition. I took `tp.system.suggestion` into the code as shown below, but it return a `template syntax error` as 
![image](https://github.com/SilentVoid13/Templater/assets/129242259/f05dbe99-1d2f-4af7-b2cf-ad0906232b85)

The codes now looks like this:  (I put the top sections as as an explanation, I copied these codes from some template I found in web, and I also do not understood what it is or,  :-)    )

```
<%*
//const head = `
//`;

//const tf = tp.file.find_tfile(tp.file.title);
//await app.vault.modify(tf, tp.file.content + head);
-%>

<%* let noteStatus;
noteStatus = tp.system.suggester(["in progress","done"],["in progress","done"], false, "file status?");

const file = tp.file.find_tfile(tp.file.title);
await app.fileManager.processFrontMatter(file, (frontmatter)=> {
	if (noteStatus === "in progress") {
		frontmatter["noteStatus"] = "in progress";
		} else (noteStatus === "done") {
			frontmatter["noteStatus"] = "done";
		}
});

const file = tp.file.find_tfile(tp.file.title);
await app.fileManager.processFrontMatter(file, (frontmatter) => {
  frontmatter["modifiedBy"] = "no body";
  frontmatter["reviewedBy"] = "no body";
  // frontmatter["noteStatus"] = `${noteStatus}`
   if (frontmatter["noteStatus"] === "done") {
	    frontmatter["finishedDate"] = tp.date.now("YYYY-MM-DD");
	    delete frontmatter["reviewedDate"];
	  } else if (frontmatter["noteStatus"] === "in progress"){
		frontmatter["reviewedDate"] = tp.date.now("YYYY-MM-DD");
		delete frontmatter["finishedDate"];
		} else {
			frontmatter["reviewedDate"] = tp.date.now("YYYY-MM-DD");
			delete frontmatter["finishedDate"];
		};
	if (frontmatter["reviewedTimes"] === undefined) {
	    frontmatter["reviewedTimes"] = 1;
		  } else {
		    frontmatter["reviewedTimes"] += 1;
		  };
})
-%>



```

Thanks again for your time and kindly helps,

Sincerely, 

FH

**Reply by Zachatoo** [reply, marked as answer]
The syntax error is because of this section.

```js
if (noteStatus === "in progress") {
  frontmatter["noteStatus"] = "in progress";
} else (noteStatus === "done") {
  frontmatter["noteStatus"] = "done";
}
```

If you're using `else`, you cannot add a condition after it in parenthesis. You have to use `else if` instead.

```js
if (noteStatus === "in progress") {
  frontmatter["noteStatus"] = "in progress";
} else if (noteStatus === "done") {
  frontmatter["noteStatus"] = "done";
}
```

Or just else, without a condition.

```js
if (noteStatus === "in progress") {
  frontmatter["noteStatus"] = "in progress";
} else {
  frontmatter["noteStatus"] = "done";
}
```

[Relevant docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/if...else).

**Reply by francisHugo** [reply]
Thanks @Zachatoo ,

I gave up to modified the codes. it is too complicated for me :-( 

Thank again.

Sincerely,

FH
